apps/events/migrations/0014_move_schedule_messages_off_periodic_tasks.py

```diff
@@ -0,0 +1,121 @@
+from datetime import datetime, timedelta
+from django.db import migrations
+import json
+
+from apps.events.models import EventAction, EventActionType, ScheduledMessage
+from apps.experiments.models import ExperimentSession
+
+def is_complete(task):
+	return datetime.now() > task.expires
+
+def calculate_next_run(task, current_time):
+	delta = timedelta(seconds=task.interval.every.total_seconds())
+	next_time = task.start_time
+	while next_time <= current_time:
+		next_time += delta
+	return next_time
+
+def get_interval_duration(interval):
+    if not interval:
+        return None
+    if interval.period == 'seconds':
+        return timedelta(seconds=interval.every)
+    elif interval.period == 'minutes':
+        return timedelta(minutes=interval.every)
+    elif interval.period == 'hours':
+        return timedelta(hours=interval.every)
+    elif interval.period == 'days':
+        return timedelta(days=interval.every)
+    elif interval.period == 'weeks':
+        return timedelta(weeks=interval.every)
+    else:
+        return None
+
+def calculate_remaining_repetitions(task, current_time):
+    interval_duration = get_interval_duration(task.interval)
+    remaining_time = task.expires - current_time if task.expires else None
+    if not remaining_time or not interval_duration:
+        return 0
+    remaining_repetitions = remaining_time / interval_duration
+    return int(remaining_repetitions)
+
+def calculate_total_triggers(task, current_time):
+    interval_duration = get_interval_duration(task.interval)
+    start_time = task.date_changed
+    if not start_time or not interval_duration:
+        return 0
+    elapsed_time = current_time - start_time
+    return int(elapsed_time / interval_duration)
+
+def calc_schedule_components(periodic_task, current_time):
+    if periodic_task.one_off:
+        schedule = periodic_task.clocked
+        frequency = 1
+        time_period = None
+        repetitions = 1
+        next_trigger_date = schedule.clocked_time
+        total_triggers = 0
+    else:
+        schedule = periodic_task.interval
+        frequency = schedule.every
+        time_period = schedule.period
+        repetitions = calculate_remaining_repetitions(periodic_task, current_time)
+        next_trigger_date = calculate_next_run(periodic_task, current_time)
+        total_triggers = calculate_total_triggers(periodic_task, current_time)
+
+    return frequency, time_period, repetitions, total_triggers, next_trigger_date
+
+
+class Migration(migrations.Migration):
+    def migrate_periodic_tasks(apps, schema_editor):
+        ScheduledMessage = apps.get_model('events', 'ScheduledMessage')
+        EventAction = apps.get_model('events', 'EventAction')
+        Experiment = apps.get_model('experiments', 'Experiment')
+        Participant = apps.get_model('experiments', 'Participant')
+        PeriodicTask = apps.get_model('django_celery_beat', 'PeriodicTask')
+        current_time = datetime.now()
```

Comment:
Ooh, careful here. `datetime.now()` returns a naive date, which is incompatable with timzone aware dates, so you will not be able to do for instance `task.expires - current_time`.

Reply:
You can use `timezone.now()` (`from django.utils import timezone`)

Reply:
[b2d92d1](https://github.com/dimagi/open-chat-studio/pull/527/commits/b2d92d17f732edd2bb9a9e62929eef7614ea1d16)